### PR TITLE
Fix #8132: managed fields failure during Restore

### LINKED ISSUE #8132: Random race condition in the restore with managed fields
**What steps did you take and what happened:**
There is a race condition in the restore operation causing random errors visible in the logs during restore operation. Below is stripped down log presenting the error during restore:

```shell
time="2024-08-15T12:21:15Z" level=error msg="error patch for managed fields test-velero: Operation cannot be fulfilled on namespaces \"test-velero\": the object has been modified; please apply your changes to the latest version and try again" logSource="/remote-source/velero/app/pkg/restore/restore.go:1682" restore=openshift-adp-2/mysql-test-velero-11abd1b8-3a6e-a161-1b3c-435a2812c5a0

[...]

time="2024-08-15T12:22:43Z" level=error msg="Cluster resource restore error: Operation cannot be fulfilled on namespaces \"test-velero\": the object has been modified; please apply your changes to the latest version and try again" logSource="/remote-source/velero/app/pkg/controller/restore_controller.go:587" restore=openshift-adp-2/mysql-test-velero-11abd1b8-3a6e-a161-1b3c-435a2812c5a0

```

**What did you expect to happen:**
Managed fields properly restored without error message in the logs.

### PR DESCRIPTION
**Pull request by mpryc**
Thank you for contributing to Velero!

# Please add a summary of your change
This commit addresses issue #8132, where an error randomly appears in the logs during the restore operation.

The error occurs due to a race condition when attempting to patch managed fields on an object that has been modified in the cluster. The error message indicates that the operation cannot be fulfilled because the object has been modified, suggesting that changes should be applied to the latest version.

To resolve this, a retry mechanism has been implemented in the restore process when encountering this error, ensuring that managed fields are properly restored without the error message appearing in the logs.

# Does your change fix a particular issue?

Fixes #8132

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [x] Updated the corresponding documentation in `site/content/docs/main`.
   - not needed

### CODE REVIEW

**Comment by anshulahuja98** [comment]
I would prefer if this is backported to 1.13 as well

**Comment by blackpiglet** [comment]
@mpryc 
I suggest to add a changelog file for this PR.

**Comment by kaovilai** [comment]
Potential for scope overlap with
https://github.com/vmware-tanzu/velero/pull/8063

**Comment by mpryc** [comment]
> @mpryc I suggest adding a changelog file for this PR.

Added, dunno how to remove `kind/changelog-not-required` label once it has been added, possibly the owner of the repo can do that.

**Comment by mpryc** [comment]
> Potential for scope overlap with #8063

I am not convinced about that. The #8063 is about retry on status update, this fix is about race condition on a particular object during restore operation and applying managed fields. It could be a general design rework of restore operation, but that's bigger chunk of work.

In short this fix addresses situation when after first time the object is created in the cluster at: https://github.com/vmware-tanzu/velero/blob/f63b71448324cf29a5366d365a1ceea900150f23/pkg/restore/restore.go#L1503

And before the patch for managed fields is being calculated at:
https://github.com/vmware-tanzu/velero/blob/f63b71448324cf29a5366d365a1ceea900150f23/pkg/restore/restore.go#L1672

There are number of operations on the in-cluster object including status update and it happens that the object being patched is not the one that represents current cluster version. This is of course done to save API calls to the cluster and we should only re-try such operation when there is real error.

I believe this is not really what you are looking into within #8063 as we are explicitly retrying on the object conflict and not other problems such as non reachable cluster API:

https://pkg.go.dev/k8s.io/client-go/util/retry#RetryOnConflict

**Comment by sseago** [comment]
@kaovilai I agree with @mpryc This seems to be a completely different retry needed. Not velero CRs -- restored resources, and not related to APIServer outages but conflict errors, so both the error matched and the backoff/duration profile is completely different.

**Comment by sseago** [comment]
@mpryc I removed the changelog-not-required label

File: changelogs/unreleased/8132-mpryc [hunk 1]
```diff
@@ -0,0 +1 @@
+Random race condition in the restore with managed fields
```
Comment (sseago) [inline changelogs/unreleased/8132-mpryc:1]: @mpryc changelog needs to have PR number, not issue number -- s/8132/8133/ in the filename.

Also, I'd rephrase the changelog to describe the fix rather than the bug (since it will appear in release notes):
"Fixed race condition for conflicts on patching managed fields" or soething like that.

Comment (mpryc) [reply changelogs/unreleased/8132-mpryc]: Changed.

**Review by sseago — APPROVED** [review]

**Comment by ywk253100** [comment]
@mpryc The patch API should not report the error mentioned in issue #8132, the reason causes #8132 is that the  `resourceVersion` field is included when doing the patch (the `resourceVersion is changed between the time window object is created and patched`). 

So a more reasonable solution would be that make sure the `resourceVersion` not being included in the patch data, e.g. make the `resourceVersion` same when compute the patch data.

**Review by ywk253100 — CHANGES_REQUESTED** [review]
See my comments

**Comment by mpryc** [comment]
> So a more reasonable solution would be that make sure the `resourceVersion` not being included in the patch data, e.g. make the `resourceVersion` same when compute the patch data.

@ywk253100 won't this have possible other issues?  There is a risk of applying a patch based on an out-of-sync version of the object. This can lead to unintended modifications, conflicts, or even overwriting of data that has been updated since the original `resourceVersion` was retrieved. Kubernetes relies on an optimistic concurrency control model in which we may be breaking by assuming we always have correct version in our cache.

**Comment by mpryc** [comment]
> @mpryc I checked the code again, seems the `resourceVersion` isn't included in the patch data, the only difference between the `withoutManagedFields` and `createdObj` is the managed fields.
> 
> ```go
> withoutManagedFields := createdObj.DeepCopy()
> createdObj.SetManagedFields(obj.GetManagedFields())
> patchBytes, err := generatePatch(withoutManagedFields, createdObj)
> ```
> 
> I'm curious why the patch operation reported the confliction error? Only patch with `resourceVersion` or update could report such error.

Any attempt to patch object in the cluster including other then `resourceVersion`  fields like status, can result in a conflict if the patch is generated from an outdated in-memory version of the object.

When the patch is applied, it may conflict with the current state of the object, leading to a conflict error. To address this, it’s important to implement a retry mechanism. Specifically defined for such scenarios, `client-go/util/retry` allows the operation to be retried using the latest version of the object when a conflict occurs.

I believe the correct fix is actually to perform retry on conflict which will ensure only this scenario is taken into consideration:

https://github.com/kubernetes/client-go/blob/02a19c375c491042890be396d94a26a69da89563/util/retry/util.go#L68-L75

**Comment by ywk253100** [comment]
> > @mpryc I checked the code again, seems the `resourceVersion` isn't included in the patch data, the only difference between the `withoutManagedFields` and `createdObj` is the managed fields.
> > ```go
> > withoutManagedFields := createdObj.DeepCopy()
> > createdObj.SetManagedFields(obj.GetManagedFields())
> > patchBytes, err := generatePatch(withoutManagedFields, createdObj)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm curious why the patch operation reported the confliction error? Only patch with `resourceVersion` or update could report such error.
> 
> @ywk253100 Any attempt to patch object in the cluster including other then `resourceVersion` fields like status, can result in a conflict if the patch is generated from an outdated in-memory version of the object.
> 
> When the patch is applied, it may conflict with the current state of the object, leading to a conflict error. To address this, it’s important to implement a retry mechanism. Specifically defined for such scenarios, `client-go/util/retry` allows the operation to be retried using the latest version of the object when a conflict occurs.
> 
> I believe the correct fix is actually to perform retry on conflict which will ensure only this scenario is taken into consideration:
> 
> https://github.com/kubernetes/client-go/blob/02a19c375c491042890be396d94a26a69da89563/util/retry/util.go#L68-L75

Hi @mpryc If you see the code I pasted, the only difference between the two objects is the `ManagedFields`, so the patch data should only contain the changed `ManagedFields` part, right? Do you mean this kind of patch is also possible to cause conflict? 
Per my understanding, it is not. Correct me if I'm wrong.
And the Kubernetes API doc says
```
Patches will never cause optimistic locking failures, and the last write will win. Patches are recommended when the full state is not read before an update, or when failing on optimistic locking is undesirable. When patching complex types, arrays and maps, how the patch is applied is defined on a per-field basis and may either replace the field's current value, or merge the contents into the current value.
```
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/

**Comment by mpryc** [comment]
> > > @mpryc I checked the code again, seems the `resourceVersion` isn't included in the patch data, the only difference between the `withoutManagedFields` and `createdObj` is the managed fields.
> > > ```go
> > > withoutManagedFields := createdObj.DeepCopy()
> > > createdObj.SetManagedFields(obj.GetManagedFields())
> > > patchBytes, err := generatePatch(withoutManagedFields, createdObj)
> > > ```
> > > 

> Hi @mpryc If you see the code I pasted, the only difference between the two objects is the `ManagedFields`, so the patch data should only contain the changed `ManagedFields` part, right? 

@ywk253100 I agree with you.  While the chances of a conflict error due to the generated patch are minimal, they are still possible, as evidenced by the errors we see in our logs.

I can't think of any other scenarios that could be causing this specific error. Implementing retries on this patch seems to be a reasonable approach to address the issue.

The  error `the object has been modified; please apply your changes to the latest version and try again
 ` in the logs is clearly linked to this line and most likely results from a race condition. Although I can't think of any other possible causes, the problem persists. If you have suggestions for a better fix, I'm open to testing them:


```
resourceClient.Patch(name, patchBytes)
```

**Comment by ywk253100** [comment]
@mpryc 
I checked the Kubernetes API server code, and seems only `PATCH` with `resourceVersion` and `PUT` API calls could cause the conflict error. If what I'm saying is right, the `PATCH` without `resourceVersion` should not cause the error.

If the `PATCH` operation could cause the conflict error, we need to add the retry logic for every call of `PATCH`. 
So let's spend a bit more time to find the root cause before merging the PR, WDYT?

BTW, in which version of Velero and Kubernetes that you see the issue? Is it reproducible?
I saw the failed resource is a namespace, is this issue specific to namespaces or not?

**Comment by mpryc** [comment]
@ywk253100 
That was visible on Velero 1.14. It's hard for me to check what the generated patch looked like and if it had the `resourceVersion` as this is random error visible post CI run in our logs, so more data is lost, but if needed I could build custom velero and collect more data for it. Would the original object, patched object, generated patch be sufficient in more debugging?

**Comment by ywk253100** [comment]
> @ywk253100 That was visible on Velero 1.14. It's hard for me to check what the generated patch looked like and if it had the `resourceVersion` as this is random error visible post CI run in our logs, so more data is lost, but if needed I could build custom velero and collect more data for it. Would the original object, patched object, generated patch be sufficient in more debugging?

@mpryc That's OK for the first step of further debugging.

BTW, seems the `managedFields` patching logic will not apply to the `namespace` resource in the normal restore process, it is skipped here https://github.com/vmware-tanzu/velero/blob/v1.14.1/pkg/restore/restore.go#L739. The only possibility the patching logic can apply to `namespace` is processing the `additionalItems` returns by the `RestoreItemAction`: https://github.com/vmware-tanzu/velero/blob/v1.14.1/pkg/restore/restore.go#L1393

Is there any `RestoreItemAction` in your deployment that returns the `namespace` as the `additionalItems`? 

Please also check whether the conflict error is reported only for `namespace` resource or not in your following debugging.

**Comment by weshayutin** [comment]
@shubham-pampattiwar please have a look for @mpryc when you have a moment.
Thank you

**Review by shubham-pampattiwar — APPROVED** [review]

**Comment by shubham-pampattiwar** [comment]
@mpryc can you please help answer @ywk253100 's concerns/questions ?

**Comment by shubham-pampattiwar** [comment]
@mpryc please resolve the conflicts and update the PR.

**Comment by mpryc** [comment]
The `Project` from the `project.openshift.io/v1` was being patched when the error was observed. Here is example Project:

```yaml
apiVersion: project.openshift.io/v1
kind: Project
metadata:
  annotations:
    openshift.io/description: ""
    openshift.io/display-name: ""
    openshift.io/requester: kube:admin
    openshift.io/sa.scc.mcs: s0:c30,c0
    openshift.io/sa.scc.supplemental-groups: 1000870000/10000
    openshift.io/sa.scc.uid-range: 1000870000/10000
    security.openshift.io/MinimallySufficientPodSecurityStandard: restricted
  creationTimestamp: "2025-10-14T13:29:47Z"
  labels:
    kubernetes.io/metadata.name: test-oadp-116
    pod-security.kubernetes.io/audit: restricted
    pod-security.kubernetes.io/audit-version: latest
    pod-security.kubernetes.io/enforce: restricted
    pod-security.kubernetes.io/enforce-version: latest
    pod-security.kubernetes.io/warn: restricted
    pod-security.kubernetes.io/warn-version: latest
  name: test-oadp-116
  resourceVersion: "11961298"
  uid: 1c396098-9d41-4d80-860d-f701bce28cec
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

**Comment by kaovilai** [comment]
Anything we need to discuss?

**Review by ywk253100 — COMMENTED** [review]
Remove my "Request changes" to unblock move forword

**Comment by ywk253100** [comment]
It's better that we can root cause this issue. 
If patching call could cause the conflict error the retrying should be applied to all patches.

**Comment by Lyndon-Li** [comment]
I discussed with @ywk253100, here are some thoughts:
We need to find the root cause and see who is the other part of the conflict:
1. If the both parts (for the conflict) are both from the same Velero restore instance, e.g., step1 create the resource->step2 update the resource->step3 update the resource again with old data. If this happens, this is a Velero restore workflow issue, a blind retry is not the best solution
2. If the other part is from some module outside of Velero restore instance, it means, someone is trying to manipulate the same resource during Velero restore, it is more like a use error. As a common sense, users should keep clear during a restore running, otherwise, no restore tool could guarantee the data are all restored as expected; and for the current issue, just as @ywk253100 mentioned, we do retry in one place, another conflict may also happen in another place.

**Comment by kaovilai** [comment]
A potential solution we also came up with internally is that the user could exclude The `Project` from the `project.openshift.io/v1` from backup because it's sync'd with Namespace object.

**Comment by mpryc** [comment]
/hold

After internal discussions there may be another solution which will not involve this change. We may drop project from the restores in the OpenShift Velero Plugin during restore, which will allow Namespace to be restored without such concurrency on managed fields.

**Comment by mpryc** [comment]
Closing as per: https://github.com/vmware-tanzu/velero/issues/8132#issuecomment-3486423976
